File: backend/app/services/kanban_service.py
"""
Service para o sistema Kanban Board.

Lógica de negócio para:
- Criação de cards com análise da IA
- Movimentação de cards (drag & drop)
- Registro de movimentos/lançamentos
- Gerenciamento de assignees e tags
"""
import json
import logging
from typing import List, Optional, Dict
from datetime import datetime
from sqlalchemy.ext.asyncio import AsyncSession

from app.models.kanban import Card, CardMovement
from app.repositories.kanban_repository import (
    CardRepository,
    CardMovementRepository,
    CardAssigneeRepository,
    CardTagRepository,
    CardColumnRepository
)
from app.services.gemini_service import gemini_service

logger = logging.getLogger(__name__)


class KanbanService:
    """Service para operações do Kanban Board"""

    def __init__(self, db: AsyncSession):
        self.db = db
        self.card_repo = CardRepository(db)
        self.movement_repo = CardMovementRepository(db)
        self.assignee_repo = CardAssigneeRepository(db)
        self.tag_repo = CardTagRepository(db)
        self.column_repo = CardColumnRepository(db)

    async def create_card_with_ai(
        self,
        company_id: int,
        user_id: int,
        title: str,
        original_text: str,
        column_id: int = 1,  # Default: Backlog
        priority: str = 'Média',
        due_date: Optional[datetime] = None
    ) -> Dict:
        """
        Cria card e analisa com IA.
        
        Returns:
            Dict com card e ai_suggestions
        """
        # Analisar com IA
        logger.info(f"Analisando card com IA: {title}")
        ai_result = await gemini_service.analyze_activity(
            title=title,
            status="Novo",
            raw_text=original_text
        )

        # Extrair dados da IA
        description = ai_result.get("description", "")
        ai_priority = ai_result.get("priority", priority)
        sub_status = ai_result.get("sub_status")
        
        # Parsear due_date se fornecido pela IA
        ai_due_date = ai_result.get("due_date")
        if ai_due_date and not due_date:
            try:
                due_date = datetime.strptime(ai_due_date, "%Y-%m-%d")
            except:
                pass

        # Criar card
        card = await self.card_repo.create(
            company_id=company_id,
            user_id=user_id,
            column_id=column_id,
            title=title,
            description=description,
            original_text=original_text,
            priority=ai_priority,
            due_date=due_date
        )

        # Atualizar SubStatus se houver
        if sub_status:
            await self.card_repo.update(
                card_id=card.CardID,
                company_id=company_id,
                SubStatus=sub_status
            )

        logger.info(f"Card {card.CardID} criado com sucesso")

        return {
            "card": card,
            "ai_suggestions": ai_result
        }

    async def save_validated_data(
        self,
        card_id: int,
        company_id: int,
        user_id: int,
        validated_data: Dict
    ) -> bool:
        """
        Salva dados validados pelo usuário após análise da IA.
        
        Cria:
        - Assignees
        - Tags
        - Movements (sub-tarefas)
        """
        try:
            # Atualizar description se fornecida
            description = validated_data.get("description")
            if description:
                await self.card_repo.update(
                    card_id=card_id,
                    company_id=company_id,
                    Description=description
                )

            # Atualizar priority se fornecida
            priority = validated_data.get("priority")
            if priority:
                await self.card_repo.update(
                    card_id=card_id,
                    company_id=company_id,
                    Priority=priority
                )

            # Atualizar sub_status se fornecido
            sub_status = validated_data.get("sub_status")
            if sub_status:
                await self.card_repo.update(
                    card_id=card_id,
                    company_id=company_id,
                    SubStatus=sub_status
                )

            # Salvar assignees
            for assignee_name in validated_data.get("assignees", []):
                await self.assignee_repo.create(
                    card_id=card_id,
                    person_name=assignee_name
                )

            # Salvar tags
            for tag_name in validated_data.get("tags", []):
                await self.tag_repo.create(
                    card_id=card_id,
                    tag_name=tag_name
                )

            # Salvar movements (sub-tarefas)
            for movement_data in validated_data.get("movements", []):
                await self.movement_repo.create(
                    card_id=card_id,
                    user_id=user_id,
                    subject=movement_data.get("subject", ""),
                    description=movement_data.get("description"),
                    time_spent=movement_data.get("estimated_time"),
                    movement_type="Update"
                )

            logger.info(f"Dados validados salvos para card {card_id}")
            return True

        except Exception as e:
            logger.error(f"Erro ao salvar dados validados: {e}")
            return False

    async def add_movement(
        self,
        card_id: int,
        company_id: int,
        user_id: int,
        subject: str,
        description: Optional[str] = None,
        time_spent: Optional[int] = None
    ) -> CardMovement:
        """
        Adiciona movimento/lançamento ao card.
        """
        # Verificar se card existe
        card = await self.card_repo.get_by_id(card_id, company_id)
        if not card:
            raise ValueError(f"Card {card_id} não encontrado")

        # Criar movimento
        movement = await self.movement_repo.create(
            card_id=card_id,
            user_id=user_id,
            subject=subject,
            description=description,
            time_spent=time_spent,
            movement_type="Update"
        )

        logger.info(f"Movimento {movement.MovementID} adicionado ao card {card_id}")
        return movement

    async def move_card(
        self,
        card_id: int,
        company_id: int,
        user_id: int,
        new_column_id: int,
        new_position: int = None
    ) -> Card:
        """
        Move card para outra coluna (drag & drop).
        Cria movimento de auditoria automaticamente.
        """
        card = await self.card_repo.move_to_column(
            card_id=card_id,
            company_id=company_id,
            new_column_id=new_column_id,
            user_id=user_id,
            new_position=new_position
        )

        if not card:
            raise ValueError(f"Card {card_id} não encontrado")

        logger.info(f"Card {card_id} movido para coluna {new_column_id}")
        return card

    async def get_card_with_details(
        self,
        card_id: int,
        company_id: int
    ) -> Optional[Dict]:
        """
        Busca card com todos os detalhes (assignees, tags, movements, etc.)
        """
        card = await self.card_repo.get_by_id(card_id, company_id)
        if not card:
            return None

        # Buscar relacionamentos
        assignees = await self.assignee_repo.get_by_card(card_id)
        tags = await self.tag_repo.get_by_card(card_id)
        movements = await self.movement_repo.get_by_card(card_id)
        total_time = await self.card_repo.get_total_time_spent(card_id)

        return {
            "card": card,
            "assignees": assignees,
            "tags": tags,
            "movements": movements,
            "total_time_spent": total_time
        }

    async def get_board(
        self,
        company_id: int
    ) -> Dict:
        """
        Retorna board completo com todas as colunas e cards.
        
        Returns:
            {
                "columns": [...],
                "cards_by_column": {
                    column_id: [cards...]
                }
            }
        """
        # Buscar colunas
        columns = await self.column_repo.get_all(company_id)

        # Buscar cards por coluna
        cards_by_column = {}
        for column in columns:
            cards = await self.card_repo.get_all(
                company_id=company_id,
                column_id=column.ColumnID
            )
            cards_by_column[column.ColumnID] = cards

        return {
            "columns": columns,
            "cards_by_column": cards_by_column
        }

    async def update_card_status(
        self,
        card_id: int,
        company_id: int,
        user_id: int,
        new_sub_status: str
    ) -> Card:
        """
        Atualiza SubStatus do card.
        Cria movimento de auditoria.
        """
        card = await self.card_repo.get_by_id(card_id, company_id)
        if not card:
            raise ValueError(f"Card {card_id} não encontrado")

        old_sub_status = card.SubStatus

        # Atualizar status
        card = await self.card_repo.update(
            card_id=card_id,
            company_id=company_id,
            SubStatus=new_sub_status
        )

        # Criar movimento de auditoria
        await self.movement_repo.create(
            card_id=card_id,
            user_id=user_id,
            subject=f"Status alterado: {old_sub_status} → {new_sub_status}",
            movement_type="StatusChange",
            old_sub_status=old_sub_status,
            new_sub_status=new_sub_status
        )

        logger.info(f"Status do card {card_id} atualizado para {new_sub_status}")
        return card

    async def complete_card(
        self,
        card_id: int,
        company_id: int,
        user_id: int
    ) -> Card:
        """
        Marca card como concluído.
        Move para coluna "Concluído" e registra data de conclusão.
        """
        # Buscar coluna "Concluído" (assumindo DisplayOrder = 5)
        columns = await self.column_repo.get_all(company_id)
        completed_column = next(
            (c for c in columns if "Conclu" in c.ColumnName),
            None
        )

        if not completed_column:
            raise ValueError("Coluna 'Concluído' não encontrada")

        # Mover para coluna concluído
        card = await self.move_card(
            card_id=card_id,
            company_id=company_id,
            user_id=user_id,
            new_column_id=completed_column.ColumnID
        )

        # Atualizar data de conclusão
        card = await self.card_repo.update(
            card_id=card_id,
            company_id=company_id,
            CompletedDate=datetime.utcnow()
        )

        logger.info(f"Card {card_id} marcado como concluído")
        return card

File: frontend/src/services/kanbanService.ts
/**
 * Service para o sistema Kanban Board
 */

import api from './api';

// ============================================================================
// Types
// ============================================================================

export interface CardColumn {
  ColumnID: number;
  ColumnName: string;
  DisplayOrder: number;
  Color: string;
  IsActive: boolean;
  CompanyID: number;
  CreatedAt: string;
}

export interface Card {
  CardID: number;
  CompanyID: number;
  UserID: number;
  ColumnID: number;
  Title: string;
  Description: string | null;
  OriginalText: string | null;
  SubStatus: string | null;
  Priority: string;
  StartDate: string | null;
  DueDate: string | null;
  CompletedDate: string | null;
  CreatedAt: string;
  IsDeleted: boolean;
}

export interface CardMovement {
  MovementID: number;
  CardID: number;
  UserID: number;
  LogDate: string;
  Subject: string;
  Description: string | null;
  TimeSpent: number | null;
  MovementType: string | null;
  OldColumnID: number | null;
  NewColumnID: number | null;
  OldSubStatus: string | null;
  NewSubStatus: string | null;
  images?: MovementImage[];
}

export interface CardAssignee {
  AssigneeID: number;
  CardID: number;
  PersonName: string;
  PersonID: number | null;
  AssignedAt: string;
}

export interface CardTag {
  CardTagID: number;
  CardID: number;
  TagName: string;
  TagColor: string | null;
  CreatedAt: string;
}

export interface CardImage {
  CardImageID: number;
  CardID: number;
  ImagePath: string;
  ImageType: string;
  Description: string;
  UploadedBy: number;
  UploadedAt: string;
}

export interface MovementImage {
  MovementImageID: number;
  MovementID: number;
  ImagePath: string;
  ImageType: string;
  Description: string;
  AIAnalysis?: string;
  UploadedAt: string;
}

export interface CardWithDetails extends Card {
  assignees: CardAssignee[];
  tags: CardTag[];
  movements: CardMovement[];
  images: CardImage[];
  total_time_spent: number | null;
}

export interface CardWithAI extends Card {
  ai_suggestions: {
    description: string;
    assignees: string[];
    systems: string[];
    tags: string[];
    priority: string;
    sub_status: string | null;
    due_date: string | null;
    movements: Array<{
      subject: string;
      description: string;
      estimated_time: number | null;
      assignee: string | null;
    }>;
  };
}

export interface Board {
  columns: CardColumn[];
  cards_by_column: Record<number, Card[]>;
}

export interface CardCreateData {
  Title: string;
  OriginalText: string;
  ColumnID?: number;
  Priority?: string;
  DueDate?: string | null;
}

export interface ValidatedCardData {
  description?: string;
  assignees?: string[];
  systems?: string[];
  tags?: string[];
  priority?: string;
  sub_status?: string | null;
  due_date?: string | null;
  movements?: Array<{
    subject: string;
    description: string;
    estimated_time: number | null;
    assignee: string | null;
  }>;
}

export interface MovementCreateData {
  Subject: string;
  Description?: string;
  TimeSpent?: number;
}

// ============================================================================
// API Calls
// ============================================================================

/**
 * Busca board completo com colunas e cards
 */
export const getBoard = async (): Promise<Board> => {
  const response = await api.get('/api/v1/kanban/board');
  return response.data;
};



/**
 * Lista colunas do board
 */
export const getColumns = async (): Promise<CardColumn[]> => {
  const response = await api.get('/api/v1/kanban/columns');
  return response.data;
};

/**
 * Cria card com análise da IA
 */
export const createCard = async (data: CardCreateData): Promise<CardWithAI> => {
  const response = await api.post('/api/v1/kanban/cards', data);
  return response.data;
};

/**
 * Lista cards
 */
export const listCards = async (
  columnId?: number,
  skip: number = 0,
  limit: number = 100
): Promise<Card[]> => {
  const params: any = { skip, limit };
  if (columnId) params.column_id = columnId;
  
  const response = await api.get('/api/v1/kanban/cards', { params });
  return response.data;
};

/**
 * Busca card com detalhes
 */
export const getCard = async (cardId: number): Promise<CardWithDetails> => {
  const response = await api.get(`/api/v1/kanban/cards/${cardId}`);
  return response.data;
};

/**
 * Atualiza card
 */
export const updateCard = async (
  cardId: number,
  data: Partial<Card>
): Promise<Card> => {
  const response = await api.put(`/api/v1/kanban/cards/${cardId}`, data);
  return response.data;
};

/**
 * Deleta card
 */
export const deleteCard = async (cardId: number): Promise<void> => {
  await api.delete(`/api/v1/kanban/cards/${cardId}`);
};

/**
 * Move card para outra coluna (drag & drop)
 */
export const moveCard = async (
  cardId: number,
  newColumnId: number,
  newPosition?: number
): Promise<Card> => {
  const response = await api.post(`/api/v1/kanban/cards/${cardId}/move`, {
    CardID: cardId,
    NewColumnID: newColumnId,
    NewDisplayOrder: newPosition
  });
  return response.data;
};

/**
 * Salva dados validados após análise da IA
 */
export const saveValidatedData = async (
  cardId: number,
  validatedData: ValidatedCardData
): Promise<void> => {
  await api.post(`/api/v1/kanban/cards/${cardId}/validate`, validatedData);
};

/**
 * Adiciona movimento/lançamento ao card
 */
export const addMovement = async (
  cardId: number,
  data: MovementCreateData
): Promise<CardMovement> => {
  const response = await api.post(`/api/v1/kanban/cards/${cardId}/movements`, data);
  return response.data;
};

/**
 * Atualiza movimento/lançamento
 */
export const updateMovement = async (
  movementId: number,
  data: MovementCreateData
): Promise<CardMovement> => {
  // Remove campos undefined/null para não enviar no request
  const cleanData = Object.fromEntries(
    Object.entries(data).filter(([_, value]) => value !== undefined && value !== null)
  );
  const response = await api.put(`/api/v1/kanban/movements/${movementId}`, cleanData);
  return response.data;
};

/**
 * Deleta movimento/lançamento
 */
export const deleteMovement = async (movementId: number): Promise<void> => {
  await api.delete(`/api/v1/kanban/movements/${movementId}`);
};

/**
 * Marca card como concluído
 */
export const completeCard = async (cardId: number): Promise<Card> => {
  const response = await api.post(`/api/v1/kanban/cards/${cardId}/complete`);
  return response.data;
};

/**
 * Adiciona tag ao card
 */
export const addTag = async (cardId: number, tagName: string): Promise<void> => {
  await api.post(`/api/v1/kanban/cards/${cardId}/tags`, {
    TagName: tagName
  });
};

/**
 * Remove tag do card
 */
export const removeTag = async (tagId: number): Promise<void> => {
  await api.delete(`/api/v1/kanban/tags/${tagId}`);
};

// ============================================================================
// Export
// ============================================================================

export const kanbanService = {
  getBoard,
  getColumns,
  createCard,
  listCards,
  getCard,
  updateCard,
  deleteCard,
  moveCard,
  saveValidatedData,
  addMovement,
  updateMovement,
  deleteMovement,
  completeCard,
  addTag,
  removeTag
};
